tests/integration/HACKING.md

```diff
@@ -0,0 +1,468 @@
+
+In some cases however, tasks have complex dependencies that are difficult to install and it may be burdensome to install them in the producer (Girder) environment. Luckily this issue can be easily ameliorated by guarding imports in your task packages:
+
+```python
+from girder_worker.app import app
+
+try:
+	from complex_dependency import complex_function
+except ImportError:
+	pass
+
+@app.task
+def some_task(*args, **kwargs):
+	return complex_function(*args, **kwargs
+
+```
+
+Producer-side code will still generate the task objects, allowing them to produce task messages,  while consumer-side code (where the complex_dependency package must be installed) will be able to execute the task body.
+
+You may additionally wish to configure setuptools using [extras_require](http://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies) to manage the different between producer and consumer dependencies:
+
+```python
+setup(
+    name="my_custom_package",
+    ...
+	install_requires=[
+		"girder_worker",
+		....
+	],
+    extras_require={
+        'consumer':  ["complex_dependency", ...],
+    }
+	...
+)
+```
+
+This means the producer side code can be installed with pip as ```my_custom_package``` and the consumer side code (that includes the complex dependencies) can be installed as ```my_custom_package[consumer]```
+
+## Communicating with Girder from inside Tasks
+
+Tasks decorated with the Girder Worker application object have two main channels of interacting with the Girder instance that produced the task. 
+
+The first is the ```job_manager``` which provides an API for interacting with the specific Girder job model that is reflecting the state of Celery Task. The job_manger is an instance of ```girder_worker.utils.JobManager```. The job\_manager is available on the task object,  which means you must decorate the task with bind equal to true (e.g. ```@app.task(bind=true)```). This means the first argument to the task will be ```self``` and the job\_manger will be available as ```self.job_manager```.  For the most part,  task state transition is handled seamlessly through Girder Worker's use of [Celery Signals](http://docs.celeryproject.org/en/latest/userguide/signals.html). This means basic state transition (e.g.  running, success, failure, retry, etc) will be handled without any intervention within the task.  If you wish to use custom states,  update the Job progress, or to log specific information within your task,  you can do so using the this job\_manager. The job_manager is automatically generated regardless of whether you use the Traditional method or the Celery method of producing your task. The job\_manager is scoped so that it only has push access to the specific job it is assigned. If your task needs to leverage other components of the Girder REST API, including pushing data back to Girder or modifying Girder collections via the API,  then you should use the ```girder_client```
```

Comment:
For the most part,  task state transition is handled seamlessly through Girder Worker's use of [Celery Signals]

What does this mean?  that Celery will take care of changing the task state?  that Celery will take care of reflecting the task state back into the Job model?  both?